Allow full customization of error unwrapping and stack tracing (and support Unwrap automatically)

bvisness opened the pull request:
This allows the Rollbar SDK to generate a full trace_chain by unwrapping errors using the newly-standardized [errors.Unwrap](https://golang.org/pkg/errors/) function. This is accomplished in a backwards-compatible way by using [golang.org/x/xerrors](https://godoc.org/golang.org/x/xerrors). When you are ok dropping support for Go versions before 1.13, you can seamlessly switch to using the standard `errors` package instead.

I pulled `Stacker` out of `CauseStacker` and made it its own interface, because `Cause` is redundant now that Go provides a standard way of wrapping errors. `CauseStacker` still exists for compatibility.

I attempted to add docs and to cover this with new tests, but I expect you may want to update or revise those.

Resolves #63.

waltjones commented:
@bvisness Thank you, and thank you for the readme cleanup as well.

I'll review as soon as possible. I need to read up on Unwrap.

mback2k commented:
Since this probably conflicts with the other existing PR #53 regarding stack trace handling, I would kindly ask you to merge the other PR first and build upon @fgblomqvist and my work if possible.

bvisness commented:
I'm re-posting my proposed design from #53 here, which should preserve backwards compatibility while allowing unwrapping and tracing for any third-party error types, even those which do not have `Cause()`, `Unwrap()`, or `Stack()`.

To sum up the overall points of the design:
- The rollbar client has an `UnwrapperFunc` and a `TracerFunc` that it can use to generate error chains and stack traces. These can be set by the user to work with any custom error type.
  - As a sub-point, the default `UnwrapperFunc` can check both `Cause()` (the old way) and `Unwrap()` (the new, standard way).
- The existing logic for error chains and stack traces is moved into default functions which are used unless the user specifies an alternative.

This design preserves existing behavior and allows the use of easy interface-based solutions for custom error types, while keeping implementation simple and allowing unwrapping and tracing to be more flexible.

```go
type Stacker interface {
    Stack() []runtime.Frame
}

type UnwrapperFunc func(error) error
type TracerFunc func(error) ([]runtime.Frame, bool)

// The client uses the following functions by default:

func DefaultUnwrapperFunc(err error) error {
    type causer interface {
        Cause() error
    }

    if c, ok := err.(causer); ok {
        return c.Cause()
    }
    if unwrapped := xerrors.Unwrap(err); unwrapped != nil {
        return unwrapped
    }
    return nil
}

func DefaultTracerFunc(err error) ([]runtime.Frame, bool) {
    if s, ok := err.(Stacker); ok {
        return s.Stack(), true
    }

    return nil, false
}

// The defaults can be overridden with the following:

func SetErrorUnwrapper(unwrapper UnwrapperFunc) {
    // set it
}

func SetStackTracer(stackTracer TracerFunc) {
    // set it
}

// transforms.go can now just use the unwrapper and stackTracer functions
// to keep the logic simpler.
```

waltjones commented:
I think you want to avoid the `xerrors` dependency in `DefaultUnwrapperFunc`. Otherwise, this looks good.

Adding the Unwrap behavior to errors/errors.go might be an option, and limit it to the pkg/errors availability so there's not a need to later transition from xerrors to errors. If someone wants to use xerrors, they can pass a custom function that uses the xerrors package.

bvisness commented:
The unwrapping behavior from `xerrors` could easily be included directly in rollbar-go to avoid the dependency. I'll go ahead and do that when I properly turn this proposal into code. Thanks for the feedback!

fgblomqvist commented:
This sounds pretty good. There are certainly other bits in xerrors/Go 1.13 that could be utilized to simplify/standardize rollbar-go's code. Is there a reason not to simply release any new rollbar-go version with 1.13 as the minimum version and just utilize the new functionality (such as Unwrap)?

mback2k commented:
@fgblomqvist I wouldn't consider Go 1.13 a good minimum default until it has matured a little.

>     * The existing logic for error chains and stack traces is moved into default functions which are used **unless the user specifies an alternative**.

@bvisness Would it make sense to allow the user to specify if his functions are alternatives or should just extend the default behavior? E.g. if the user just wants to use his functions and/or also keep the default behavior at the same time to handle all kinds of error types.

bvisness commented:
@mback2k I was imagining that the default functions would be exported so you could call them from your custom functions. That way you can either extend the default behavior or override it completely. Maybe that’s more granular than we need though, since you’ll probably only be setting your own unwrapper/tracer to get your specific flavor of errors to work. 

I agree that having Go 1.13 as a minimum dependency would probably be undesirable for a lot of people, especially since at my company alone (which stays pretty up to date) we still have a bunch of projects lingering as far back as 1.10 because the changes to dependency management are a pain.

bvisness commented on errors/go.mod:
```diff
@@ -1,3 +1,5 @@
 module github.com/rollbar/rollbar-go/errors
 
+go 1.13
+
```
This was added automatically since I'm using Go 1.13, but it shouldn't prevent older versions of Go from using this module. See https://golang.org/doc/go1.12#modules.

bvisness commented on rollbar_test.go:
```diff
@@ -402,58 +405,120 @@ func (cs cs) Stack() []runtime.Frame {
+		if effect.Stack()[0] != getOrBuildFrames(effect, effect2, 0, DefaultStackTracer)[0] {
+			t.Error("should use stack from child, not parent")
+		}
+	})
```
This test is equivalent to the old `TestGetOrBuildStackOfCauseStackerWithParent`, with one important change. In the original, the code was as follows:

```go
cause := fmt.Errorf("cause")
effect := cs{fmt.Errorf("effect"), cause, getCallersFrames(0)}
effect2 := cs{fmt.Errorf("effect2"), effect, getCallersFrames(0)}
if effect2.Stack()[0] != getOrBuildFrames(effect2, effect, 0, nil)[0] {
	t.Error("should use stack from effect2")
}
```

You can see that `effect2` wraps `effect`, which wraps `cause`. So `effect2` is the parent of `effect`. But, `getOrBuildFrames` passed `effect2` as the child and `effect` as the parent.

Since I believe this to be an incorrect test, I have swapped `effect2` and `effect` and modified the test to assert that it yields the stack from `effect`.

bvisness replied:
It's worth noting that I don't think this actually changes anything about the logic tested by this case. Both `effect` and `effect2` are traceable, which means `getOrBuildFrames` would just return the stack from whichever one was passed as the child.

bvisness commented:
Ok, I think this is ready for code review. All the existing tests pass (after some small modifications to accommodate the new signatures, and with one exception I have commented on), and I have added new tests for the new stuff.

You can see that I chose to go with the approach laid out in my proposal regarding extending vs. overriding. That approach grants users the most flexibility and is simpler to implement anyway. I have added documentation and examples to encourage users to include the `Default` things in their custom functions.

Looking forward to your feedback!

waltjones commented:
@bvisness This looks ready to merge to me. Curious whether @mback2k has any comments though.

mback2k commented:
@bvisness @waltjones LGTM. Thanks!
